readers/nonforwarding: don't emit partition_end on next_partition,fast_forward_to

gusev-p opened the pull request:
The series fixes the `make_nonforwardable` reader, it shouldn't emit `partition_end` for previous partition after `next_partition()` and `fast_forward_to()`

Fixes: #12249

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4580/

denesb commented on readers/mutation_reader.cc:
```diff
@@ -385,11 +385,6 @@ flat_mutation_reader_v2::~flat_mutation_reader_v2() {
-void flat_mutation_reader_v2::impl::forward_buffer_to(const position_in_partition& pos) {
-    clear_buffer();
-    _buffer_size = compute_buffer_size(*_schema, _buffer);
-}
```
This smells like a bug to me. This method was supposed to clear buffer *up to* the specified position.
Did we drop this on the floor with the v2 migration? @tgrabiec?

tgrabiec replied:
It was needed with flat_mutation_reader which had range_tombstone fragments. They could be relevant to the next position_range on fast_forward_to() because they could hold information which extends past the current range. With flat_mutation_reader_v2 we no longer have such fragments, so mutation_fragment::relevant_for_range_assuming_after() was dropped and this is effectively clearing all. We just forgot to remove it. So I think this change is fine.

denesb commented on readers/mutation_readers.cc:
```diff
@@ -411,15 +411,19 @@ flat_mutation_reader_v2 make_nonforwardable(flat_mutation_reader_v2 r, bool sing
         flat_mutation_reader_v2 _underlying;
         bool _single_partition;
         bool _static_row_done = false;
+        bool _partition_is_open = false;
```
Can't we reuse the existing `_static_row_done` flag (maybe rename it to `_partition_header_done`)? The underlying forwarding reader will always emit the partition header (partition start and static row if any) first, then require fast forwarding to proceed.
I'm fine with 2 flags too if you think it helps code clarity.

gusev-p replied:
> Can't we reuse the existing `_static_row_done` flag (maybe rename it to `_partition_header_done`)? The underlying forwarding reader will always emit the partition header (partition start and static row if any) first, then require fast forwarding to proceed.
> I'm fine with 2 flags too if you think it helps code clarity.

Reader should emit a `partition_end` only if there was a `partition_start` before. `_static_row_done` only tells us that the call to `fast_forward_to` was made, is says nothing about the content of the stream, it could be empty.

denesb commented on test/boost/flat_mutation_reader_test.cc:
```diff
@@ -727,6 +728,33 @@ SEASTAR_TEST_CASE(test_make_forwardable) {
     });
 }
 
+SEASTAR_TEST_CASE(test_make_nonforwardable) {
```
We need test names that hint at what they are testing.

This test seems to stress the emitting logic for partition_end, so this should be in the name somehow.

gusev-p replied:
This test is not only about `partition_end`, it contains all the basic checks for `make_nonforwardable`, there were no tests for this function before. Each check is a separate region with its own comment. This way it's easy to reuse various helpers and test data. We can separate each check into its own test, but I don't see much value in it. The already existing `test_make_forwardable` is structured in the same way.

denesb replied:
We should add a test which subject this reader to the mutation source reader torture suite.

denesb replied:
That suite would have cough all these problems

gusev-p replied:
> We should add a test which subject this reader to the mutation source reader torture suite.

How to do it? Could you share an example?

denesb replied:
Search for `run_mutation_source_tests` in `test/boost/mutation_reader_test.cc` or `test/boost/flat_mutation_reader_test.cc` and you will see plenty of examples.
The gist is this:
* You create an underlying reader with the passed-in mutations. The most convenient is to create a memtable, dump the mutations into them, then create a memtable reader. You will always have to create this with `streamed_mutation::forwarding::yes`.
* Wrap this reader with the nonforwarding reader.
* When the test asks for forwarding reader, you can use `make_forwarding()` to make your nonforwarding reader forwarding one.

tgrabiec replied:
This reader can't work with the general mutation_source tests, because it works only with a single partition. mutation source tests assume that the mutation source can accept multiple partitions.

tgrabiec replied:
We could extract a battery of single-partition tests out of the general mutation source battery of tests.

gusev-p replied:
I added `test_make_nonforwardable_from_mutations_as_mutation_source`

denesb reviewed (changes requested):
I don't understand the first commit at all. It seems to do some hacky workaround over a hack where fragments are injected into the middle of the reader stack. I think if this is causing problems, the injection should be fixed so that it happens on the bottom-most reader instead. Or maybe I completely misunderstood.

The rest looks good to me, with the exception of the last commit, which looks like some mistake made in the migration.

The PR title needs to be more specific: "fix the bug" is way too generic, it should be something like:
"readers/nonforwarding: don't emit partition_end on next_partition,fast_forward_to"

gusev-p commented:
> I don't understand the first commit at all. It seems to do some hacky workaround over a hack where fragments are injected into the middle of the reader stack. I think if this is causing problems, the injection should be fixed so that it happens on the bottom-most reader instead. Or maybe I completely misunderstood.

Did you read the commit message?

denesb commented:
> Did you read the commit message?

Yes, multiple times even.

denesb commented:
> > Did you read the commit message?
> 
> Yes, multiple times even.

Looks like not enough times. I read it even and now I remembered that the cache reader does this trick where it pops off the partition start from the disk reader, because the content of the disk reader is injected into the middle of an already started partition. This is how the cache reader weaves together the cache and disk.
Let me have another look at the code too.

gusev-p commented:
> I read it once again and now I remembered that the cache reader does this trick where it pops off the partition start from the disk reader

Yes, that's the point.

denesb commented:
Ok, the first commit makes sense now, but @tgrabiec should review too. Cache is tricky.

gusev-p commented:
The test failure seems related to https://github.com/scylladb/scylla-jmx/issues/194, not sure how to deal with it.

denesb commented:
> The test failure seems related to [scylladb/scylla-jmx#194](https://github.com/scylladb/scylla-jmx/issues/194), not sure how to deal with it.

I don't see anything related to JMX in the test results. It looks like a timeout while trying to replace a node.

gusev-p commented:
> > The test failure seems related to [scylladb/scylla-jmx#194](https://github.com/scylladb/scylla-jmx/issues/194), not sure how to deal with it.
> 
> I don't see anything related to JMX in the test results. It looks like a timeout while trying to replace a node.

Ah, I ran it locally, had some JMX issue, and though it was the same on the CI.

tgrabiec commented on row_cache.cc:
```diff
@@ -637,12 +633,8 @@ class scanning_and_populating_reader final : public flat_mutation_reader_v2::imp
-                }
                 return make_ready_future<flat_mutation_reader_v2_opt>(std::move(fropt));
             } else {
                 _secondary_in_progress = false;
```
> However, this makes a problem:
> now we pop partition_start from the underlying reader
> in autoupdating_underlying_reader::move_to_next_partition
> and manually push it back to downstream readers
> bypassing nonforwardable reader. This means if we
> change the logic in nonforwardable reader as described
> we will end up with partition_start without partition_end
> in the downstream readers.

I don't think make_nonforwardable() should rely on this assumption, that is that it is invoked on a stream which was never consumed from. It's enough if make_nonforwardable()::reader::next_partition() sets a flag which puts the reader at EOS without emitting anything.

tgrabiec replied:
But the patch is good as a cleanup.

gusev-p replied:
> It's enough if make_nonforwardable()::reader::next_partition() sets a flag which puts the reader at EOS without emitting anything.

It's true for the case when `single_partition=true`, and this indeed should be fixed/tested. In general, there may be other partitions and we should handle that if `make_nonforwardable` pretends to be generic. In particular, we have to handle situations when this next partition is empty or not empty. In the first case we shouldn't emit `partition_end`, in the second case we should. The `_partition_is_open` flag was added to distinguish these two cases. Here we strive to preserve the semantics of intra-partition fast-forwarding mode described in [your comment](https://github.com/scylladb/scylladb/issues/12249#issuecomment-1439061442).

tgrabiec commented:
The cover letter should mention "Fixes #..."

gusev-p commented:
v2:
  * fix `next_partition` and `fast_forward_to` when `single_partition` is false;
  * use `SEASTAR_THREAD_TEST_CASE`;
  * add `test_make_nonforwardable_from_mutations_as_mutation_source` which exercises `make_nonforwardable` with the help of `run_mutation_source_tests`.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4623/

gusev-p commented:
v3:
  * rebased on master

gusev-p commented:
@denesb @tgrabiec ping

denesb commented on test/boost/flat_mutation_reader_test.cc:
```diff
@@ -727,6 +728,33 @@ SEASTAR_TEST_CASE(test_make_forwardable) {
 }
 
+SEASTAR_TEST_CASE(test_make_nonforwardable) {
+    return seastar::async([] {
```
Should use `SEASTAR_THREAD_TEST_CASE`.

denesb commented on test/boost/flat_mutation_reader_test.cc:
```diff
@@ -728,124 +728,122 @@ SEASTAR_TEST_CASE(test_make_forwardable) {
-        simple_schema s;
-        tests::reader_concurrency_semaphore_wrapper semaphore;
-        const auto permit = semaphore.make_permit();
+SEASTAR_THREAD_TEST_CASE(test_make_nonforwardable) {
```
This should not be a separate commit, instead the change should be retroactively applied to the past commits.

denesb commented on test/boost/flat_mutation_reader_test.cc:
```diff
@@ -846,6 +846,33 @@ SEASTAR_THREAD_TEST_CASE(test_make_nonforwardable) {
+                range,
+                slice,
+                streamed_mutation::forwarding::yes);
+            reader = make_nonforwardable(std::move(reader), false);
```
Instead of `false` you should pass `muts.size() == 1` (after the squashing).

gusev-p replied:
done

denesb reviewed (approved):
LGTM, only some minor comments.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4631/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4635/

gusev-p commented:
v4:
  * `test_make_nonforwardable_from_mutations_as_mutation_source` revealed a problem in `make_forwardable::reader::next_partition`, added a fix for it;
  * `flat_mutation_reader_test` cleanup, `seastar::async -> SEASTAR_THREAD_TEST_CASE`;
  * `test_make_nonforwardable_from_mutations_as_mutation_source`, added `single_partition = squashed_muts.size() == 1;` condition to check the logic when `single_partition==true`;
  * `make_nonforwardable::reader::next_partition` adjusted condition when `_single_partition==true`, set `EOS` only if `_partition_is_open==true`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4666/

gusev-p commented:
@denesb ping, new version (hope the last one), tests passed
